File: Engine/Source/Graphics/RenderTexture.cpp
#include <spdlog/spdlog.h>
#include <Yonai/Graphics/RenderTexture.hpp>

using namespace glm;
using namespace Yonai;
using namespace Yonai::Graphics;

bool Yonai::Graphics::IsDepthFormat(TextureFormat format)
{
	switch (format)
	{
	default: return false;
	case TextureFormat::Depth: return true;
	}
}

RenderTexture::RenderTexture(ivec2 resolution, TextureFormat format, unsigned int samples) : m_ID(0)
{
	m_Format = format;
	m_Samples = samples;
	m_Resolution = resolution;

	Recreate();
}

RenderTexture::~RenderTexture() {}

void RenderTexture::SetResolution(ivec2 resolution)
{
	if (m_Resolution == resolution)
		return;

	m_Resolution = resolution;
	Recreate();
}

void RenderTexture::SetSamples(unsigned int samples)
{
	if (samples == m_Samples)
		return;

	m_Samples = samples;
	Recreate();
}

void RenderTexture::SetFormat(TextureFormat format)
{
	if (format == m_Format)
		return;

	// Create in new format
	m_Format = format;
	Recreate();
}

unsigned int RenderTexture::GetID() { return m_ID; }
ivec2 RenderTexture::GetResolution() { return m_Resolution; }
TextureFormat RenderTexture::GetFormat() { return m_Format; }
unsigned int RenderTexture::GetSamples() { return m_Samples; }

void RenderTexture::Recreate()
{
	if (IsDepthFormat(m_Format))
		CreateDepthTexture();
	else
		CreateColourTexture();
}

void RenderTexture::Bind(unsigned int textureIndex)
{
}

void RenderTexture::Unbind() {}

void RenderTexture::CreateDepthTexture() {}

void RenderTexture::CreateColourTexture() {}

void RenderTexture::CopyTo(RenderTexture* destination)
{
	ivec2 outputResolution =
	{
		min(m_Resolution.x, destination->m_Resolution.x),
		min(m_Resolution.y, destination->m_Resolution.y)
	};

	
}

#pragma region Internal Calls
#include <Yonai/Scripting/InternalCalls.hpp>

ADD_MANAGED_METHOD(RenderTexture, Load, void*, (glm::ivec2* resolution, unsigned char textureFormat, unsigned int samples), Yonai.Graphics)
{ return new RenderTexture(*resolution, (TextureFormat)textureFormat, samples); }

ADD_MANAGED_METHOD(RenderTexture, Destroy, void, (void* handle), Yonai.Graphics)
{ if(handle) delete handle; }

ADD_MANAGED_METHOD(RenderTexture, Bind, void, (void* handle, unsigned int index), Yonai.Graphics)
{ ((RenderTexture*)handle)->Bind(index); }

ADD_MANAGED_METHOD(RenderTexture, Unbind, void, (void* handle), Yonai.Graphics)
{ ((RenderTexture*)handle)->Unbind(); }

ADD_MANAGED_METHOD(RenderTexture, SetFormat, void, (void* handle, unsigned char format), Yonai.Graphics)
{ ((RenderTexture*)handle)->SetFormat((TextureFormat)format); }

ADD_MANAGED_METHOD(RenderTexture, SetSamples, void, (void* handle, unsigned int samples), Yonai.Graphics)
{ ((RenderTexture*)handle)->SetSamples(samples); }

ADD_MANAGED_METHOD(RenderTexture, SetResolution, void, (void* handle, glm::ivec2* resolution), Yonai.Graphics)
{ ((RenderTexture*)handle)->SetResolution(*resolution); }

ADD_MANAGED_METHOD(RenderTexture, GetFormat, unsigned char, (void* handle), Yonai.Graphics)
{ return (unsigned char)((RenderTexture*)handle)->GetFormat(); }

ADD_MANAGED_METHOD(RenderTexture, GetSamples, unsigned int, (void* handle), Yonai.Graphics)
{ return ((RenderTexture*)handle)->GetSamples(); }

ADD_MANAGED_METHOD(RenderTexture, GetResolution, void, (void* handle, glm::ivec2* resolution), Yonai.Graphics)
{ *resolution = ((RenderTexture*)handle)->GetResolution(); }

ADD_MANAGED_METHOD(RenderTexture, CopyTo, void, (void* handle, void* destination), Yonai.Graphics)
{ ((RenderTexture*)handle)->CopyTo((RenderTexture*)destination); }

#pragma endregion

File: Apps/AquaScriptCore/Graphics/RenderTexture.cs
using System;
using System.Runtime.CompilerServices;

[assembly: InternalsVisibleTo("AquaScriptEditor")]
namespace AquaEngine.Graphics
{
	public class RenderTexture : IDisposable
	{
		private IVector2 m_Resolution = IVector2.Zero;
		public IVector2 Resolution
		{
			get => m_Resolution;
			set
			{
				m_Resolution = value;
				_SetResolution(Handle, ref m_Resolution);
			}
		}

		private uint m_Samples;
		public uint Samples
		{
			get => m_Samples;
			set => _SetSamples(Handle, m_Samples = value);
		}

		private TextureFormat m_Format = TextureFormat.None;
		public TextureFormat Format
		{
			get => m_Format;
			set => _SetFormat(Handle, (byte)(m_Format = value));
		}

		internal IntPtr Handle { get; private set; }
		private bool m_CreatedResource = false;

		public RenderTexture(IVector2 resolution, TextureFormat format = TextureFormat.RGBA8, uint samples = 1)
		{
			m_Resolution = resolution;
			m_Format = format;
			m_Samples = samples;
			m_CreatedResource = true;

			Handle = _Load(
				ref m_Resolution,
				(byte)m_Format,
				m_Samples
				);
		}

		internal RenderTexture(IntPtr handle)
		{
			Handle = handle;
			m_CreatedResource = false;

			m_Format = (TextureFormat)_GetFormat(Handle);
			m_Samples = _GetSamples(Handle);
			_GetResolution(Handle, ref m_Resolution);
		}

		public void Dispose()
		{
			Log.Warning("RenderTexture disposed");

			// if _Load is called memory is allocated, _Destroy frees that memory.
			// if RenderTexture(IntPtr) constructor is called instead, we don't manage this memory.
			if (m_CreatedResource && Handle != IntPtr.Zero)
			{
				_Destroy(Handle);
				Handle = IntPtr.Zero;
			}
		}

		public void Bind(uint textureIndex = 0) => _Bind(Handle, textureIndex);
		public void Unbind() => _Unbind(Handle);

		public void CopyTo(RenderTexture destination)
		{
			if (destination != null)
				_CopyTo(Handle, destination.Handle);
		}

		#region Internal Calls
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static IntPtr _Load(ref IVector2 resolution, byte textureFormat, uint samples);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _Destroy(IntPtr handle);

		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _Bind(IntPtr handle, uint index);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _Unbind(IntPtr handle);

		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _CopyTo(IntPtr handle, IntPtr destination);

		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _SetFormat(IntPtr handle, byte format);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _SetSamples(IntPtr handle, uint samples);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _SetResolution(IntPtr handle, ref IVector2 resolution);

		[MethodImpl(MethodImplOptions.InternalCall)] private extern static byte _GetFormat(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static uint _GetSamples(IntPtr handle);
		[MethodImpl(MethodImplOptions.InternalCall)] private extern static void _GetResolution(IntPtr handle, ref IVector2 resolution);
		#endregion
	}
}
